jenkins-plugin/src/main/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepExecution.java

```diff
@@ -644,21 +650,43 @@ private String setupSettingFile(@Nonnull Collection<Credentials> credentials) th
 
             if ((settings = ws.child(settingsPath)).exists()) {
                 // settings file residing on the agent
-                console.format("[withMaven] use Maven settings provided on the build agent '%s' %n", settingsPath);
-                LOGGER.log(Level.FINE, "Copying maven settings file from build agent {0} to {1}", new Object[]{settings, settingsDest});
+                if (LOGGER.isLoggable(Level.FINE)) {
+                    console.format("[withMaven] using Maven settings provided on the build agent '%s' %n", settingsPath);
+                    LOGGER.log(Level.FINE, "Copying maven settings file from build agent {0} to {1}", new Object[] { settings, settingsDest });
+                }
                 settings.copyTo(settingsDest);
+                envOverride.put("MVN_SETTINGS", settingsDest.getRemote());
+                return settingsDest.getRemote();
             } else {
                 throw new AbortException("Could not find file '" + settings + "' on the build agent");
             }
-            envOverride.put("MVN_SETTINGS", settingsDest.getRemote());
-            return settingsDest.getRemote();
         }
 
-        // Settings provided by the global maven configuration
-        SettingsProvider settingsProvider = GlobalMavenConfig.get().getSettingsProvider();
+        SettingsProvider settingsProvider;
+
+        MavenConfigFolderOverrideProperty overrideProperty = getMavenConfigOverrideProperty();
+        StringBuilder mavenSettingsLog=new StringBuilder();
+
+        if (overrideProperty != null && overrideProperty.getSettings() != null) {

```

Comment:
So for instance here, you don't even use `isOverride`?

Reply:
see the getter above, it will return null if false.